daemon/hosts.go

```diff
@@ -0,0 +1,166 @@
+package daemon // import "github.com/docker/docker/daemon"
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes/docker"
+	hostconfig "github.com/containerd/containerd/remotes/docker/config"
+	"github.com/docker/docker/registry"
+	"github.com/pkg/errors"
+)
+
+const (
+	defaultPath = "/v2"
+)
+
+// RegistryHosts returns the registry hosts configuration for the host component
+// of a distribution image reference.
+func (daemon *Daemon) RegistryHosts(host string) ([]docker.RegistryHost, error) {
+	hosts, err := hostconfig.ConfigureHosts(context.Background(), hostconfig.HostOptions{
+		// TODO: Also check containerd path when updating containerd to use multiple host directories
+		HostDir: hostconfig.HostDirFromRoot(registry.CertsDir()),
+	})(host)
+	if err == nil {
+		// Merge in legacy configuration if provided
+		if cfg := daemon.Config(); len(cfg.Mirrors) > 0 || len(cfg.InsecureRegistries) > 0 {
+			hosts, err = daemon.mergeLegacyConfig(host, hosts)
+		}
+	}
+
+	return hosts, err
+}
+
+func (daemon *Daemon) mergeLegacyConfig(host string, hosts []docker.RegistryHost) ([]docker.RegistryHost, error) {
+	// If no hosts provided, nothing to do.
+	// If multiple hosts provided, then a mirror configuration is already provided and
+	// should not overwrite with legacy config.
+	if len(hosts) == 0 || len(hosts) > 1 {
+		return hosts, nil
+	}
+	sc := daemon.registryService.ServiceConfig()
+	if host == "docker.io" && len(sc.Mirrors) > 0 {
+		var mirrorHosts []docker.RegistryHost
+		for _, mirror := range sc.Mirrors {
+			h := hosts[0]
+			h.Capabilities = docker.HostCapabilityPull | docker.HostCapabilityResolve
+
+			u, err := url.Parse(mirror)
+			if err != nil || u.Host == "" {
+				u, err = url.Parse(fmt.Sprintf("//%s", mirror))
+			}
+			if err == nil && u.Host != "" {
+				h.Host = u.Host
+				h.Path = strings.TrimRight(u.Path, "/")
+				if !strings.HasSuffix(h.Path, defaultPath) {
```

Comment:
This part is confusing (but maybe I need to dig in more to grasp); we check if the path _ends_ with default path, and if not we _prepend_ the default path. Maybe it could use a short comment

Reply:
Updated this logic to _mostly_ match the original mirror logic. Seems the original mirror logic didn't check the suffix just unconditionally added "/v2". Updated to add at the end.

There have been different ideas around what the provided endpoint path should mean, some want it as a repository subpath and others as a root prefix. Here just going to add at the end now and mentioned in a comment since this is just for compatibility with the existing mirror strings. Using the new host configuration from containerd allows much flexibility for users to specify whatever they want without us trying to override what they set.